docs/api_overview.md

```diff
@@ -0,0 +1,472 @@
+`Context`. For example:
+
+```c++
+static RegisterContextFactory register_ExampleContext(
+    CONTEXT_FACTORY(ExampleContext), ROOT_FACTORY(ExampleRootContext));
+```
+
+`ROOT_FACTORY` and `CONTEXT_FACTORY` are convenience macros for lambdas that
+create instances of the specified `RootContext` and `Context` subclasses.
+        
+## `RootContext`
+
+`RootContext` instances are associated with the plugin as a whole. Accordingly,
+they receive callbacks corresponding to different events in the lifecycle of a
+plugin:
+
+* `validateConfiguration`: May be called to validate the configuration data that
+  will be passed to a plugin on startup, e.g. in the control plane.
+* `onCreate`: called when context is created.
+* `onStart`: called on plugin start.
+* `onConfigure`: called on plugin start, and any time configuration subsequently
+  changes.
+* `onDone`: called when the plugin is being shut down.
+* `onDelete`: called after the plugin has acknowledged shutdown is complete, as
+  indication to release resources.
+
+`RootContexts` can also receive other callbacks for events that are not strictly
+tied to an incoming stream:
+
+* `onTick`: called when a timer fires. See [Timers].
+* `onQueueReady`: called when data arrives on a SharedQueue. See [Shared
+  queues].
+
+`RootContext` defines methods for initiating outbound HTTP and gRPC callouts:
+
+* `httpCall`: initiates an HTTP callout (i.e. outbound HTTP request). See [HTTP
+  callouts].
+* `grpcSimpleCall`: initiates a gRPC callout (i.e. outbound gRPC call). See
+  [gRPC callouts].
+* `grpcCallHandler`: initiates a gRPC callout (i.e. outbound gRPC call). See
+  [gRPC callouts].
+* `grpcStreamHandler`: initiates a streaming gRPC callout (i.e. outbound gRPC
+  call). See [gRPC callouts].
+
+For API details, see doc comments for the `RootContext` class in
+[proxy_wasm_api.h].
+
+## `Context`
+
+`Context` instances are associated with individual streams--i.e. incoming HTTP
+or gRPC requests, or TCP streams if the network proxy is acting as an L4
+proxy. `Context` instances receive callbacks corresponding to different events
+in the lifecycle of a stream:
+
+* `onCreate`: called when handling of a new stream starts.
+* `onDone`: called when the host is done processing the stream.
+* `onLog`: called after the host is done processing the stream, if the plugin is
+  being used for access logging (for example, in Envoy, as a
+  [WasmAccessLog](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/access_loggers/wasm/v3/wasm.proto#extensions-access-loggers-wasm-v3-wasmaccesslog)
+  extension).
+* `onDelete`: called after the plugin has completed all processing related to
+  the stream, as indication to release resources.
+
+`Context` instances also receive callbacks corresponding to stream events. For
+HTTP or gRPC streams, these are:
+
+* `onRequestHeaders`: called when HTTP or gRPC request headers are received.
+* `onRequestBody`: called when HTTP or gRPC request body data is received.
```

Comment:
For the BODY handlers, say something about chunking? e.g. add "(for each chunk)"

Reply:
+1 to document chunking, because the callback names are not suggesting that.

Reply:
Done